### PR TITLE
[FEATURE] Introduce Text Roles

### PR DESCRIPTION
**Pull request by linawolf**
resolves https://github.com/doctrine/rst-parser/issues/191

### CODE REVIEW

File: lib/Environment.php [hunk 1]
```diff
@@ -53,6 +54,9 @@ class Environment
     /** @var Reference[] */
     private $references = [];
 
+    /** @var TextRole[] */
```
Comment (greg0ire) [inline lib/Environment.php]: Let's document the key type as well
```suggestion
    /** @var array<string, TextRole> */
```

File: lib/Parser.php [hunk 2]
```diff
@@ -88,6 +89,15 @@ public function initDirectives(): void
         }
     }
 
+    public function initTextRoles(): void
+    {
+        $textRoles = $this->kernel->getTextRoles();
+
+        foreach ($textRoles as $textRole) {
```
Comment (greg0ire) [inline lib/Parser.php]: Let's inline this.

```suggestion
        foreach ($this->kernel->getTextRoles() as $textRole) {
```

File: lib/Environment.php (hunk 1 above)
Comment (linawolf) [reply lib/Environment.php]: Someone ™️ should do the same for the references and directives...

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]
Approved, but is there a reason why `lib/TextRoles/TextRole.php` is an abstract class and not an interface?

**Comment by linawolf** [comment]
> Approved, but is there a reason why `lib/TextRoles/TextRole.php` is an abstract class and not an interface?

My idea was to implement it the same way like abstract class Directive and abstract class Reference. So fututure changes would be less breaking.

**Review by wouterj — APPROVED** [review]
Looking good to me.

We currently use references for our custom text roles, but this seems much more clean. In the future, we might want to implement the reference feature of this component as a text role, but let's keep the changes and impact low for now :)

**Comment by greg0ire** [comment]
Thanks @linawolf

**Pull request merged**